File: app/src/main/java/com/stedi/randomimagegenerator/app/other/ChainSerializable.java
package com.stedi.randomimagegenerator.app.other;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.Serializable;

public class ChainSerializable implements Serializable {
    private ChainSerializable next;

    private final Serializable state;

    public ChainSerializable(@NonNull Serializable state) {
        this.state = state;
    }

    @NonNull
    public ChainSerializable createNext(@NonNull Serializable state) {
        next = new ChainSerializable(state);
        return next;
    }

    @Nullable
    public ChainSerializable getNext() {
        return next;
    }

    @NonNull
    public Serializable getState() {
        return state;
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/other/ChainSerializable.kt
package com.stedi.randomimagegenerator.app.other

import java.io.Serializable

class ChainSerializable(private val state: Serializable?) : Serializable {
    private var next: ChainSerializable? = null

    fun addChain(state: Serializable?) = ChainSerializable(state).apply { this@ChainSerializable.next = this }

    fun getChain() = next

    fun get() = state
}